### PR TITLE
Refactor the reports

### PR DESCRIPTION
**Pull request by vmarkovtsev**
Signed-off-by: Vadim Markovtsev <vadim@sourced.tech>

### CODE REVIEW

File: lookout/style/format/tests/test_quality_report.py [hunk 1]
```diff
@@ -69,39 +71,88 @@ def __exit__(self, *args):
...
+                          language=self.language, model_path=self.model_path,
+                          config={"uast_break_check": False, "aggregate": True})
+        qcount = output.count(
+            "# Quality report for javascript / <unknown url> <unknown reference> <unknown commit>")
+        self.assertEqual(qcount, 1)
+        output = "\n".join(output)
+        output = output[:output.find("# Model report for <unknown url>")]
+        metrics = _get_precision_recall_f1_support(output)
+        self.assertEqual(
+            metrics, (0.7443157321054488, 0.7700059276822763, 0.7323340980410543, 3374))
```
Comment (zurk) [inline lookout/style/format/tests/test_quality_report.py]: can this numbers be changed suddenly after some changes in the code?

File: lookout/style/format/tests/test_quality_report.py [hunk 2]
```diff
@@ -69,39 +71,88 @@ def __exit__(self, *args):
...
             with tempfile.NamedTemporaryFile() as empty_model:
                 with self.assertRaises(ValueError):
-                    quality_report(
+                    print_reports(
                         input_pattern=input_pattern, bblfsh=self.bblfsh, language=self.language,
                         model_path=empty_model, config={"uast_break_check": False})
 
     @unittest.skipUnless(os.getenv("LONG_TESTS", False),
-                         "Timeis labeled and unfiltere-consuming tests are skipped by default.")
+                         "Time-consuming tests are skipped by default.")
```
Comment (zurk) [inline lookout/style/format/tests/test_quality_report.py]: oops.

File: lookout/style/format/benchmarks/top_repos_quality.py [hunk 3]
```diff
@@ -173,8 +172,8 @@ def wrapped_capture_quality_report(*args, **kwargs):
     try:
         QualityReportAnalyzer.generate_model_report = \
             capture_report(QualityReportAnalyzer.generate_model_report, "model")
-        QualityReportAnalyzer.generate_report = \
-            capture_report(QualityReportAnalyzer.generate_report, "quality")
+        QualityReportAnalyzer.generate_quality_report = \
+            capture_report(QualityReportAnalyzer.generate_quality_report, "quality")
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply lookout/style/format/benchmarks/top_repos_quality.py]: It used to be `generate_report` and `generate_model_report`, which is much more confusing. How do you suggest to rename this one?

[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply lookout/style/format/benchmarks/top_repos_quality.py]: After all, what the function really does is it prints confusion matrices and quality metrics, so "quality" is not far from the truth. Maybe we should change the docstring instead.

File: lookout/style/format/benchmarks/top_repos_quality.py [hunk 4]
```diff
@@ -214,29 +213,27 @@ def calc_avg(arr: Sequence[Sequence], col: int) -> float:
...
-    # TODO: append a JSON codeblock to each report and parse it with json.loads
-    # See https://github.com/src-d/style-analyzer/pull/326/files#r236667952
-    for line in report.splitlines():
-        if "weighted avg" in line:
-            _, prec, recall, f1, support = line.split("` | `")
-            prec, recall, f1, support = [float(f.replace("`", "").replace("|", "")) for f in
-                                              (prec, recall, f1, support)]
-            return prec, recall, f1, int(support)
+    data = _get_json_data(report)["weighted avg"]
+    return data["precision"], data["recall"], data["f1-score"], data["support"]
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply lookout/style/format/benchmarks/top_repos_quality.py]: I don't want to touch upstream code at this point

File: lookout/style/format/general_report.py [hunk 5]
```diff
@@ -83,12 +87,13 @@ def generate_report(vnodes, n_files):
             file_mispred.append(vn.path)
     file_stat = Counter(file_mispred)
     to_show = file_stat.most_common()
-    if n_files > 0:
-        to_show = to_show[:n_files]
+    if max_files > 0:
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply lookout/style/format/general_report.py]: Does not improve here I think

File: lookout/style/format/general_report.py [hunk 6]
```diff
@@ -103,38 +108,20 @@ def generate_model_report(model: FormatModel,
...
-        rules_report.append(template.render(
-            rules_desc=rules_desc, min_support=min_support, max_support=max_support,
-            min_conf=min_conf, max_conf=max_conf, language=language, rules=rules,
-            packages=packages,
-        ))
-    return "\n".join(rules_report)
+            raise NotFittedError(language)
+    template = _load_jinja2_template("model_report.md.jinja2")
+    return template.render(model=model, languages=languages, FeatureExtractor=FeatureExtractor,
+                           describe_rule=describe_rule)
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply lookout/style/format/general_report.py]: At the same time, it is super explicit and actually very close to Golang ideas

File: lookout/style/format/general_report.py [hunk 7]
```diff
@@ -307,14 +294,17 @@ def generate_model_report(self) -> str:
...
 
         :return: report.
         """
-        return generate_report(chain.from_iterable(fix.all_vnodes for fix in fixes),
-                               self.config["n_files"])
+        fixes = list(fixes)
+        vnodes = chain.from_iterable(fix.all_vnodes for fix in fixes)
+        # FIXME(vmarkovtsev): we are taking the first fix here which does not work for >1 language
+        return generate_quality_report(
+            fixes[0].language, self.model.ptr, vnodes, self.config["max_files"])
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply lookout/style/format/general_report.py]: Good point

File: lookout/style/format/templates/model_report.md.jinja2 [hunk 8]
```diff
@@ -1,20 +1,50 @@
...
-`{{ "Min/max support: %s/%s, min/max conf: %s/%s" % (min_support, max_support, min_conf,
-max_conf) }}`
+```
+model.dump()
+```
+
+{% set data = {} %}
+{% for language in languages %}
+## {{ language }}
+    {% set scns = namespace(min_support=1000000000000, max_support=-1, min_conf=1, max_conf=0) %}
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply lookout/style/format/templates/model_report.md.jinja2]: Unfortunately no, Jinja2 does not have `float` defined. I guess such a huge number will always work :D

[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply lookout/style/format/templates/model_report.md.jinja2]: I also tried `1e64` and it doesn't work, too

File: lookout/style/format/templates/model_report.md.jinja2 [hunk 9]
```diff
@@ -1,20 +1,50 @@
...
-| `{{package[0]}}` | {{package[1]}} |
-{% endfor %}
\ No newline at end of file
+{% endfor %}{# languages #}
+<details>
+    <summary>Machine-readable report</summary>
+```json
+{{ data | tojson }}
+```
+</details>
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply lookout/style/format/templates/model_report.md.jinja2]: The number of new lines should be as low a possible, but not lower

File: lookout/style/format/templates/quality_report.md.jinja2 [hunk 10]
```diff
@@ -1,13 +1,15 @@
...
 
-### Classification report:
 | class | precision | recall | f1-score | support |
 |----:|:-----|:-------|:-----|:-----|
 {% for class in cl_report %}
 | `{{ class }}` | `{{ "%.2f" % cl_report[class]["precision"] }}` | `{{ "%.2f" % cl_report[class]["recall"] }}` | `{{ "%.2f" % cl_report[class]["f1-score"] }}` | `{{ cl_report[class]["support"] }}` |
 {% endfor %}
 
-### Confusion matrix:
+### Confusion matrix
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply lookout/style/format/templates/quality_report.md.jinja2]: That's Egor's legacy... Fixed

File: lookout/style/format/templates/quality_report.md.jinja2 [hunk 11]
```diff
@@ -16,9 +18,17 @@
...
-{% endfor %}
\ No newline at end of file
+{% endfor %}
+
+<details>
+    <summary>Machine-readable report</summary>
+```json
+{{ cl_report | tojson }}
+```
+</details>
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply lookout/style/format/templates/quality_report.md.jinja2]: Now I got you

File: lookout/style/format/tests/test_quality_report.py (hunk 1 above)
Comment (vmarkovtsev) [reply lookout/style/format/tests/test_quality_report.py]: Yes, so this is a good indicator of such changes. Let's see if we have to change them frequently, leaving for now

File: lookout/style/format/benchmarks/top_repos_quality.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply lookout/style/format/benchmarks/top_repos_quality.py]: `generate_quality_report` does not use the model, I checked it

[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply lookout/style/format/benchmarks/top_repos_quality.py]: So do you propose `generate_dataset_report`?

File: lookout/style/format/benchmarks/general_report.py [hunk 12]
```diff
@@ -307,14 +294,19 @@ def generate_model_report(self) -> str:
...
-    def generate_report(self, fixes: Iterable[FixData]) -> str:
+    def generate_quality_report(self, fixes: Iterable[FixData]) -> str:
         """
         Generate quality report: classification report, confusion matrix, files with most errors.
 
         :return: report.
         """
-        return generate_report(chain.from_iterable(fix.all_vnodes for fix in fixes),
-                               self.config["n_files"])
+        fixes = list(fixes)
```
Comment (zurk) [inline lookout/style/format/benchmarks/general_report.py]: Why do we need it? Remove it.

Comment (zurk) [reply lookout/style/format/benchmarks/general_report.py]: It can be memory-intence

Comment (vmarkovtsev) [reply lookout/style/format/benchmarks/general_report.py]: We need it because we cannot take the first element from an `Iterable`

**Pull request merged**